src/Tools/dotnet-gcdump/DotNetHeapDump/EventPipeDotNetHeapDumper.cs

```diff
@@ -0,0 +1,238 @@
+                readerTask = Task.Factory.StartNew(delegate
+                {
+                    gcDumpSession = new EventPipeSession(processID, new List<Provider> { new Provider("Microsoft-Windows-DotNETRuntime", (ulong)(ClrTraceEventParser.Keywords.GCHeapSnapshot)) });
+                    int gcNum = -1;
+
+                    // Start the providers and trigger the GCs.  
+                    log.WriteLine("{0,5:n1}s: Requesting a .NET Heap Dump", sw.Elapsed.TotalSeconds);
+
+                    gcDumpSession.Source.Clr.GCStart += delegate (GCStartTraceData data)
+                    {
+                        if (data.ProcessID != processID)
+                        {
+                            return;
+                        }
+
+                        etwDataPresent = true;
+
+                        if (gcNum < 0 && data.Depth == 2 && data.Type != GCType.BackgroundGC)
+                        {
+                            gcNum = data.Count;
+                            log.WriteLine("{0,5:n1}s: Dot Net Dump Started...", sw.Elapsed.TotalSeconds);
+                        }
+                    };
+
+                    gcDumpSession.Source.Clr.GCStop += delegate (GCEndTraceData data)
+                    {
+                        if (data.ProcessID != processID)
+                        {
+                            return;
+                        }
+
+                        if (data.Count == gcNum)
+                        {
+                            log.WriteLine("{0,5:n1}s: DotNet GC Complete.", sw.Elapsed.TotalSeconds);
+                            dumpComplete = true;
+                        }
+                    };
+
+                    gcDumpSession.Source.Clr.GCBulkNode += delegate (GCBulkNodeTraceData data)
+                    {
+                        if (data.ProcessID != processID)
+                        {
+                            return;
+                        }
+
+                        etwDataPresent = true;
+
+                        if ((sw.Elapsed - lastEtwUpdate).TotalMilliseconds > 500)
+                        {
+                            log.WriteLine("{0,5:n1}s: Making  GC Heap Progress...", sw.Elapsed.TotalSeconds);
+                        }
+
+                        lastEtwUpdate = sw.Elapsed;
+                    };
+
+                    if (memoryGraph != null)
+                    {
+                        dumper.SetupCallbacks(memoryGraph, gcDumpSession.Source, processID.ToString());
+                    }
+
+                    listening = true;
+                    gcDumpSession.Source.Process();
+                    log.WriteLine("{0,5:n1}s: EventPipe Listener dying", sw.Elapsed.TotalSeconds);
+                });
+
+                // Wait for thread above to start listening (should be very fast)
+                while (!listening)
+                {
+                    readerTask.Wait(1);
```

Comment:
I think this only serves to ensure the initial log message in the preceding task? If you pulled that log message out of the task I don't think wait would be necessary.

Reply:
This wait also ensure that the EventPipe session has started and all the callbacks have been registered before we start timing for timeout.

Reply:
I don't understand why we need to ensure that has occurred?

Reply:
Hm, now that I've re-read it again, you're right.  I can't see any reason why we should wait here.